src/charm.py

```diff
@@ -58,16 +59,65 @@ def __init__(self, *args):
         self.framework.observe(self.on.update_status, self._on_update_status)
         self.framework.observe(self.on.get_primary_action, self._on_get_primary_action)
 
+        # if a new leader has been elected, reconfigure the replica set
+        self.framework.observe(self.on.leader_elected, self._on_leader_elected)
+
+    def _on_leader_elected(self, event: ops.charm.LeaderElectedEvent) -> None:
+        """Handles leader elected event in the case that a new leader has elected.
+
+        In the case that a new leader has been elected the new leader reconfigures the replica set.
+
+        Args:
+            event: The triggering leader elected event
+        """
+        # only reconfigure if previous leader stepped down without reconfiguring
+        # relation data only accepts string, hence "False"
+        if self._peers.data[self.app].get("_new_leader_must_reconfigure", "False") == "False":
+            return
+
+        # wait to reconfigure until a primary has been elected
+        if not self._primary:
+            event.defer()
+            self.unit.status = WaitingStatus("waiting to reconfigure replica set")
+            return
+
+        self._reconfigure(event)
+
+        # reset need for new leader to reconfigure
+        self._peers.data[self.app]["_new_leader_must_reconfigure"] = "False"
+
     def _on_mongodb_relation_departed(self, event: ops.charm.RelationDepartedEvent) -> None:
         """Removes the unit from the MongoDB replica set.
 
         Args:
             event: The triggering relation departed event.
         """
+        # acquire removed unit
+        # TODO update this to use ops framework after this issue is addressed:
+        # https://github.com/canonical/operator/issues/707
+        # this will be updated in the following PR
+        departing_unit = os.environ.get("JUJU_DEPARTING_UNIT")
+
+        # If primary is leaving, allow it to step down before reconfiguring
+        if departing_unit == self.unit.name and departing_unit == self._primary:
```

Comment:
one small nit: it is more readable to assign a variable for conditionals - e.g.

```
is_departing_primary_unit = departing_unit == self.unit.name and departing_unit == self._primary
if is_departing_primary_unit:
    ...
```

with the added variable, it takes a lot less time to determine what we are conditioning on. thoughts?

Reply:
Hmm I actually find this scheme harder to ready, but maybe that is just what I am used to 🤷 

But I am not very opinionated on these issues, since there is a comment above I will keep it as is, unless you feel strongly :)